designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala

```diff
@@ -0,0 +1,149 @@
+class DeploymentService(
+    scenarioMetadataRepository: ScenarioMetadataRepository,
+    deploymentRepository: DeploymentRepository,
+    // TODO: we shouldn't call legacy service, instead we should new service from the legacy one
+    legacyDeploymentService: LegacyDeploymentService,
+    dbioRunner: DBIOActionRunner,
+    clock: Clock
+)(implicit ec: ExecutionContext) {
+
+  def processCommand(
+      command: DeploymentCommand
+  ): Future[Either[RunDeploymentError, Unit]] =
+    command match {
+      case command: RunDeploymentCommand =>
+        runDeployment(command)
+    }
+
+  private def runDeployment(
+      command: RunDeploymentCommand
+  ): Future[Either[RunDeploymentError, Unit]] =
+    dbioRunner.run(
+      (for {
+        scenarioMetadata <- getScenarioMetadata(command)
+        _ <- EitherT.cond[DB](
+          command.user.can(scenarioMetadata.processCategory, Permission.Deploy),
+          (),
+          NoPermissionError
+        )
+        _         <- saveDeployment(command, scenarioMetadata)
+        runResult <- invokeLegacyRunDeploymentLogic(command, scenarioMetadata)
+      } yield runResult).value
+    )
+
+  private def getScenarioMetadata(command: RunDeploymentCommand) =
+    EitherT[DB, RunDeploymentError, ProcessEntityData](
+      scenarioMetadataRepository
+        .getScenarioMetadata(command.scenarioName)
+        .map(_.map(Right(_)).getOrElse(Left(ScenarioNotFoundError(command.scenarioName))))
+    )
+
+  private def saveDeployment(command: RunDeploymentCommand, scenarioMetadata: ProcessEntityData) =
+    EitherT(
+      deploymentRepository.saveDeployment(
+        DeploymentEntityData(command.id, scenarioMetadata.id, Timestamp.from(clock.instant()), command.user.id)
+      )
+    ).leftMap(e => ConflictingDeploymentIdError(e.id))
+
+  private def invokeLegacyRunDeploymentLogic(command: RunDeploymentCommand, scenarioMetadata: ProcessEntityData) =
+    EitherT[DB, RunDeploymentError, Unit](
+      DB.from(
+        // TODO: Currently it doesn't use our deploymentId. Instead, it uses action id
+        legacyDeploymentService
+          .processCommand(
+            LegacyRunDeploymentCommand(
+              CommonCommandData(
+                ProcessIdWithName(scenarioMetadata.id, command.scenarioName),
+                command.comment,
+                command.user
+              ),
+              savepointPath = None,
+              nodesDeploymentData = command.nodesDeploymentData,
+            )
+          )
+          .transform { result =>
+            result
+              .map[Either[RunDeploymentError, Unit]](_ => Right(()))
+              .recover { case CommentValidationError(msg) =>
+                Left(NewCommentValidationError(msg))
+              }
+          }
+      )
+    )
+
+  def getDeploymentStatus(
+      id: DeploymentId
+  )(implicit loggedUser: LoggedUser): Future[Either[GetDeploymentStatusError, StatusName]] =
+    dbioRunner.run(
```

Comment:
in this transaction, I see that only getting the deployment ID is a "real" DBIO action. Other things (permission check and getting scenario status) are lifted from Future. So maybe it's better to run the first thing and then "work" on the Futures level?